theano/sandbox/gpuarray/basic_ops.py

```diff
@@ -19,27 +21,83 @@
 except ImportError:
     pass
 
-from .type import GpuArrayType
+from .type import GpuArrayType, GpuArrayConstant, gpu_context_type, get_context
 from .fp16_help import write_w
 
 
-def as_gpuarray_variable(x):
-    if getattr(x, 'owner', None):
-        if isinstance(x.owner.op, HostFromGpu):
-            return x.owner.inputs[0]
-        elif (isinstance(x.owner.op, GpuFromHost) and
-              x.owner.inputs[0].owner and
-              isinstance(x.owner.inputs[0].owner.op, HostFromGpu)):
-            return x.owner.inputs[0].owner.inputs[0]
+def as_gpuarray_variable(x, context_name):
+    # If this is already some form of variable, try to avoid an extra transfer
+    if isinstance(x, Variable):
+        while True:
+            # If we are already a GpuArrayVariable in the right context
+            # then there is nothing to do.
+            if (isinstance(x.type, GpuArrayType) and
+                    x.type.context_name == context_name):
+                return x
+
+            # If x is the result of a transfer, try to dig through.
+            if getattr(x, 'owner', None):
+                if isinstance(x.owner.op, HostFromGpu):
+                    x = x.owner.inputs[0]
+                    continue
+                if isinstance(x.owner.op, GpuFromHost):
+                    x = x.owner.inputs[0]
+                    continue
+                if isinstance(x.owner.op, GpuToGpu):
+                    x = x.owner.inputs[0]
+                    continue
+
+            # If none of the conditions where met, then continue with
+            # the rest of the body
+            break
+
+        # If we couldn't deal with transfers, then maybe it's a tensor
+        if isinstance(x.type, tensor.TensorType):
```

Comment:
Why special case TensorType? The generic version should work. If not, add a comment to tell why.